File: 백준/Gold/22251. 빌런 호석/빌런 호석.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int p = Integer.parseInt(st.nextToken());
        int x = Integer.parseInt(st.nextToken());

        HashMap<Integer, int[]> map = new HashMap<>();

        map.put(0, new int[] {1,1,1,0,1,1,1});
        map.put(1, new int[] {0,0,1,0,0,1,0});
        map.put(2, new int[] {1,0,1,1,1,0,1});
        map.put(3, new int[] {1,0,1,1,0,1,1});
        map.put(4, new int[] {0,1,1,1,0,1,0});
        map.put(5, new int[] {1,1,0,1,0,1,1});
        map.put(6, new int[] {1,1,0,1,1,1,1});
        map.put(7, new int[] {1,0,1,0,0,1,0});
        map.put(8, new int[] {1,1,1,1,1,1,1});
        map.put(9, new int[] {1,1,1,1,0,1,1});

        int[][] dp = new int[10][10];

        for (int i = 0 ; i < 10; i++) {
            for (int j = i+1; j < 10; j++) {
                int cnt = 0;
                for (int l = 0; l < 7; l++) {
                    if (map.get(i)[l] != map.get(j)[l]) {
                        cnt += 1;
                    }
                }
                dp[i][j] = cnt;
                dp[j][i] = cnt;
            }
        }

        int cnt = 0;
        String stringX = "0".repeat(k - String.valueOf(x).length()) + String.valueOf(x);

        for (int i = 1; i < n+1; i++) {
            int zero = k - String.valueOf(i).length();
            String cur = "0".repeat(zero) + i;
            int total = 0;

            if (cur.equals(stringX)) {
                continue;
            }

            for (int j = 0; j < k; j++) {
                total += dp[cur.charAt(j) - '0'][stringX.charAt(j) - '0'];

                if (total > p) {
                    break;
                }
            }

            if (total <= p) {
                cnt +=1;
            }
        }

        System.out.println(cnt);

    }
}

File: 백준/Gold/22251. 빌런 호석/빌런 호석.py
# 빌런 호석
import sys
input = sys.stdin.readline

n, k, p, x = map(int,input().split())
dic = {0:[1,1,1,0,1,1,1],
       1:[0,0,1,0,0,1,0],
       2:[1,0,1,1,1,0,1],
       3:[1,0,1,1,0,1,1],
       4:[0,1,1,1,0,1,0],
       5:[1,1,0,1,0,1,1],
       6:[1,1,0,1,1,1,1],
       7:[1,0,1,0,0,1,0],
       8:[1,1,1,1,1,1,1],
       9:[1,1,1,1,0,1,1]
       }

dp = [[0] * 10 for _ in range(10)] # dp[n][m] = n에서 m까지 가는데 필요한 갯수

change = [[0] * 10 for _ in range(10)]

for i in range(10):
    for j in range(i+1, 10):
        cnt = 0
        for k in range(7):
            if dic[i][k] != dic[j][k]:
                cnt += 1
        dp[i][j] = cnt
        dp[j][i] = cnt

cnt = 0
string_x = (k - len(str(x))) * "0" + str(x)

for i in range(1, n+1):
    zero = k - len(str(i))
    cur = zero*"0" + str(i)
    total = 0

    if cur == string_x:
        continue
    
    for j in range(k):
        total += dp[int(cur[j])][int(string_x[j])]

        if total > p:
            break

    if total <= p:
        cnt += 1

print(cnt)
